### PR TITLE
feat(plugin-image-gallery): Image Grid (PE-53, PE-54)

### PR DESCRIPTION
**Pull request by shn-srl**
**Linear Issues**:
- PE-54
- PE-53

**Demo Links**: 
- [Editor Preview](https://frontend-git-pe-54-plugin-image-gallery-serlo.vercel.app/___editor_preview?state={"plugin"%3A"rows"%2C"state"%3A[{"plugin"%3A"imageGallery"%2C"state"%3A{"images"%3A[{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fwww.shutterstock.com%2Fimage-vector%2Ftwo-opponents-facing-each-other-260nw-2296619399.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"4f3872a4-0900-43fc-bb02-85c218320616"}}%2C"id"%3A"07ae3472-4b40-4a8c-8605-ee3efedabf69"}%2C{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fc6.staticflickr.com%2F9%2F8520%2F28357073053_cafcb3da6f_b.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"a251f2ed-a184-45ff-b025-1d47267dbccd"}}%2C"id"%3A"d934acaa-6f8f-4701-934a-aeeb1daabd3c"}%2C{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fc8.staticflickr.com%2F9%2F8104%2F28973555735_ae7c208970_b.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"6c47127b-d410-46ef-a7e3-a2692766dd33"}}%2C"id"%3A"16fd4402-9b52-4238-a174-d8b4c1180688"}%2C{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fcdn.pixabay.com%2Fphoto%2F2024%2F02%2F15%2F14%2F31%2Fdonkey-8575524_960_720.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"f8612633-273a-4ae0-a2f4-04e4da3ecc4c"}}%2C"id"%3A"2a264899-87a6-46fd-84be-067c4bbc0291"}]%2C"orderedIds"%3A""}%2C"id"%3A"c1efe02e-9554-4122-ba26-7667234411a1"}]})
- [Create Article](https://frontend-git-pe-54-plugin-image-gallery-serlo.vercel.app/entity/create/Article/1377)

**To do:**

- [x] Bug: Uploading more than two images makes the plugin break
- [x] Loading state for image upload
- [x] Single image upload

### CODE REVIEW

**Comment by elbotho** [comment]
Thanks for the demo Links, really helpful here. Some initial feedback:

**Lightbox**

- tall images are clipping for me 
- when opening an image it's not selected and no toolbar is visible (only after click)
- will there be a lightbox for the static view at some point?


image
image

--- 

**Bundle size**

The bundle size increase comes from the gallery plugin I guess? I think especially for the static view this is a lot of js just for the layout. I'm aware that there will be sorting later and that can be tricky to implement.

image

Personally I'd love a simple css based solution for the grid and then using some library or code based on react-dnd only inside the editing view for sorting. (e.g. like [react-easy-sort](https://github.com/ValentinH/react-easy-sort) might be worth a look). But I can also not say with confidence how much work this way would actually be.

At least we should load the component dynamically.

**Comment by shn-srl** [comment]
> Personally I'd love a simple css based solution for the grid and then using some library or code based on react-dnd only inside the editing view for sorting. (e.g. like [react-easy-sort](https://github.com/ValentinH/react-easy-sort) might be worth a look). But I can also not say with confidence how much work this way would actually be.
> 
> At least we should load the component dynamically.

I have a branch with the sorting functionality done with react-dnd. Can look at a css only solution for the grid



> 
> **Lightbox**
> 
> 
>     * will there be a lightbox for the static view at some point?

This is in another issue: https://linear.app/serlo/issue/PE-57/lightbox-desktop

**Comment by shn-srl** [comment]
> Thanks for the demo Links, really helpful here. Some initial feedback:
> 
> **Lightbox**
> 
>     * tall images are clipping for me
> 
>     * when opening an image it's not selected and no toolbar is visible (only after click)
> 
>     * will there be a lightbox for the static view at some point?

This is in another issue:

**Comment by shn-srl** [comment]
@elbotho I have fixed the clipping (as far as I can tell), and the toolbar visibility.

Removed the react-photo-album library and implemented a solution with tailwind and styles

File: packages/editor/src/plugins/image-gallery/index.ts [hunk 1]
```diff
@@ -7,17 +7,15 @@ import {
...
 
 function createImageGalleryState() {
   return object({
-    images: list(
-      child({
-        plugin: EditorPluginType.Image,
-      })
-    ),
+    images: list(child({ plugin: EditorPluginType.Image })),
+    orderedIds: string(''),
```
Comment (elbotho) [inline packages/editor/src/plugins/image-gallery/index.ts]: do we need `orderedIds` in the editor state? 
I thought the (ordered) array would have this information already.

**Review by elbotho — COMMENTED** [review]
> @elbotho I have fixed the clipping (as far as I can tell), and the toolbar visibility.
> 
> Removed the react-photo-album library and implemented a solution with tailwind and styles

Thank you for the adaptations!
I added another clipping fix, it seems it only happened on smaller screens (vertical space).

Bundle size still goes crazy, but maybe that's because of the CI being weird. I build it locally and it seemed fine.

I like the simpler layout approach 👍
Something [like this](https://play.tailwindcss.com/tNkRUbIsum) would also work nicely, but also has more dom elements. Not sure.

**Comment by shn-srl** [comment]
> > @elbotho I have fixed the clipping (as far as I can tell), and the toolbar visibility.
> > Removed the react-photo-album library and implemented a solution with tailwind and styles
> 
> Thank you for the adaptations! I added another clipping fix, it seems it only happened on smaller screens (vertical space).
> 
> Bundle size still goes crazy, but maybe that's because of the CI being weird. I build it locally and it seemed fine.
> 
> I like the simpler layout approach 👍 Something [like this](https://play.tailwindcss.com/tNkRUbIsum) would also work nicely, but also has more dom elements. Not sure.

image

Nice, thanks for the change.

I see the bundles size reduced.

Removed the orderedIds

File: packages/editor/src/plugins/image-gallery/index.ts (hunk 1 above)
Comment (shn-srl) [reply packages/editor/src/plugins/image-gallery/index.ts]: Removed

**Comment by elbotho** [comment]
this is the most relevant line:
image

**Pull request merged**